### PR TITLE
[FuncOutline] Fix function outlining to skip linalg.fill/linalg.skip

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit fixes function outlining to skip linalg.fill/linalg.skip
-- Without this `linalg::isElementwise(LinalgOp)` seems to be qualifying both as elementwise ops.

Signed-off-by: Abhishek Varma <abhvarma@amd.com>

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp [hunk 1]
```diff
@@ -88,6 +88,8 @@ void AMDAIELinalgFunctionOutliningPass::runOnOperation() {
   DenseMap<Operation *, std::string> computeOpToOutlinedFuncMap;
   SmallVector<Operation *> toBeErased;
   moduleOp.walk([&](linalg::LinalgOp computeOp) {
+    if (isa<linalg::CopyOp, linalg::FillOp>(computeOp))
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp:91]: This is related to my suggestion in https://github.com/nod-ai/iree-amd-aie/pull/862#discussion_r1820847047 : suggestion there is to make sure we handle each op correctly with fail until we explicitly categorized it as 'must-outline' or 'must-not-outline'.

Could you have have 

auto mustOutline = [](Operation * op) { return isa<linalg::MatmulOp>(op); }
auto mustNotOutline = [](Operation * op) { return isa<linalg::CopyOp, linalg::FillOp>(op); }

And then 

```
if (mustNotOutline(computeOp)) return WalkResult::skip();
else if (!mustOutine(computeOp)) { // path to signal pass failure that the op is uncategorized; }
else { // continue as usual }
```

Above suggestion is not a robust way to do outlining, but at least we'll know exactly what is getting outlined and what isn't!

**Comment by yzhang93** [comment]
Can you report the matmul performance w/wo this PR?

**Comment by newling** [comment]
@Abhishek-Varma  If you want to measure time (and you don't already know what I'm about to say..) you can add `<chrono>` to measure time around here: https://github.com/nod-ai/iree-amd-aie/blob/9b0582f3cc7e04b9f1747e115e2ee43225ab77f0/runtime/src/iree-amd-aie/driver/xrt-lite/direct_command_buffer.cc#L193

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp [hunk 2]
```diff
@@ -88,6 +100,12 @@ void AMDAIELinalgFunctionOutliningPass::runOnOperation() {
   DenseMap<Operation *, std::string> computeOpToOutlinedFuncMap;
   SmallVector<Operation *> toBeErased;
   moduleOp.walk([&](linalg::LinalgOp computeOp) {
+    if (mustNotOutline(computeOp))
+      return WalkResult::skip();
+    else if (!mustOutline(computeOp)) {
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp:105]:
```suggestion
    if (mustNotOutline(computeOp)) {
      return WalkResult::skip();
    } else if (!mustOutline(computeOp)) {
```
See style guide: https://google.github.io/styleguide/cppguide.html#Formatting_Looping_Branching

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/linalg_function_outlining.mlir [hunk 3]
```diff
@@ -181,3 +158,52 @@ func.func @elemwise_example(%A: memref<4xf32>, %C: memref<4xbf16>, %B: memref<4x
...
+      %2 = arith.extf %in_17 : bf16 to f32
+      %3 = arith.mulf %1, %2 : f32
+      %4 = arith.addf %out, %3 : f32
+      %5 = arith.addf %4, %4 : f32
+      linalg.yield %5  : f32
+    }
+    amdaie.end
+  }
+  return
+}
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/linalg_function_outlining.mlir:209]:
```suggestion
}

```
Add new line.

**Review by jtuyls — APPROVED** [review]
LGTM, just a few nits.

**Review by newling — APPROVED** [review]
LGTM, thanks for the changes

**Pull request merged**